include/envoy/server/filter_config.h

```diff
@@ -0,0 +1,192 @@
+   * @return AccessLogManager for use by the entire server.
+   */
+  virtual AccessLog::AccessLogManager& accessLogManager() PURE;
+
+  /**
+   * @return Upstream::ClusterManager& singleton for use by the entire server.
+   */
+  virtual Upstream::ClusterManager& clusterManager() PURE;
+
+  /**
+   * @return Event::Dispatcher& the main thread's dispatcher. This dispatcher should be used
+   *         for all singleton processing.
+   */
+  virtual Event::Dispatcher& dispatcher() PURE;
+
+  /**
+   * @return DrainManager& singleton for use by the entire server.
+   */
+  virtual DrainManager& drainManager() PURE;
+
+  /**
+   * @return whether external healthchecks are currently failed or not.
+   */
+  virtual bool healthCheckFailed() PURE;
+
+  /**
+   * @return the server-wide http tracer.
+   */
+  virtual Tracing::HttpTracer& httpTracer() PURE;
+
+  /**
+   * @return the server's init manager. This can be used for extensions that need to initialize
+   *         after cluster manager init but before the server starts listening. All extensions
+   *         should register themselves during configuration load. initialize() will be called on
+   *         each registered target after cluster manager init but before the server starts
+   *         listening. Once all targets have initialized and invoked their callbacks, the server
+   *         will start listening.
+   */
+  virtual Init::Manager& initManager() PURE;
+
+  /**
+   * @return information about the local environment the server is running in.
+   */
+  virtual const LocalInfo::LocalInfo& localInfo() PURE;
+
+  /**
+   * @return RandomGenerator& the random generator for the server.
+   */
+  virtual Envoy::Runtime::RandomGenerator& random() PURE;
+
+  /**
+   * @return a new ratelimit client. The implementation depends on the configuration of the server.
+   */
+  virtual RateLimit::ClientPtr
+  rateLimitClient(const Optional<std::chrono::milliseconds>& timeout) PURE;
+
+  /**
+   * @return Runtime::Loader& the singleton runtime loader for the server.
+   */
+  virtual Envoy::Runtime::Loader& runtime() PURE;
+
+  /**
+   * DEPRECATED: Do not call this function. It will be removed post 1.4.0 and is needed for other
+   * deprecated functionality.
+   */
+  virtual Instance& server() PURE;
+
+  /**
+   * @return Stats::Scope& the filter's stats scope.
+   */
+  virtual Stats::Scope& scope() PURE;
+
+  /**
+   * @return ThreadLocal::Instance& the thread local storage engine for the server. This is used to
+   *         allow runtime lockless updates to configuration, etc. across multiple threads.
+   */
+  virtual ThreadLocal::Instance& threadLocal() PURE;
+};
+
+enum class NetworkFilterType { Read, Write, Both };
+
+/**
+ * This lambda is used to wrap the creation of a network filter chain for new connections as they
+ * come in. Filter factories create the lambda at configuration initialization time, and then they
+ * are used at runtime. This maps JSON -> runtime configuration.
```

Comment:
It's a bit unclear from this type signature that this is what is happening. All you get is a `Network::FilterManager`, whereas when I look at how this is used later, it's capturing JSON config in the lambda in some examples. I think maybe this should be made clearer.